### PR TITLE
Support roundingType and outputSizes options for pool2d.

### LINKED ISSUE #135: Support 'roundingType' and 'outputSizes' options for pool2d operations
Align to latest change of pool2d operations: https://github.com/webmachinelearning/webnn/pull/208

### PR DESCRIPTION
**Pull request by BruceDai**
Fix #135.
Depend on https://github.com/tensorflow/tfjs/pull/5849.

@huningxin PTAL, thanks.

### CODE REVIEW

**Comment by huningxin** [comment]
Hi @BruceDai , thanks for the PR. You mentions it depends on a tfjs enhancement. May I know what is the gap? roundingType or outputSizes or both of them? If there is no gap to support roundingType. I would recommend to focus on that first without depending on the tfjs release.

**Comment by BruceDai** [comment]
Hi @huningxin.

Currently [our pool2d implementation's invoked `tf.pool()` function](https://github.com/webmachinelearning/webnn-polyfill/blob/master/src/nn/ops/pool2d.ts#L114) doesn't support `dimRoundingMode` attributes which new option `roundingType` could be mapped to, and `tf.pool()` function either doesn't support `outputSizes` attribute, so I submitted the PR to enable/support these two attributes for tfjs pool.

 If we invoked `tf.avgPool()` or `tf.maxPool()` functions to replace `tf.pool()`, we could support new option `roundingType` firstly.  Thanks.

**Comment by huningxin** [comment]
> If we invoked `tf.avgPool()` or `tf.maxPool()` functions to replace `tf.pool()`, we could support new option `roundingType` firstly. Thanks.

It sounds good to me.

**Comment by BruceDai** [comment]
> If we invoked tf.avgPool() or tf.maxPool() functions to replace tf.pool(), we could support new option roundingType firstly. 

@huningxin I'm sorry to deliver the wrong message,  I missed considering the case of `dilations`. `tf.avgPool()` and `tf.maxPool()` functions could not support `dilations` directly, while tf.pool() function could do pre-processing with `dilations`, as below code, then pass that to tf.avgPool()` and `tf.maxPool()` functions
```ts
function pool_<T extends Tensor3D|Tensor4D>(
    input: T|TensorLike, windowShape: [number, number]|number,
    poolingType: 'avg'|'max',
    pad: 'valid'|'same'|number|conv_util.ExplicitPadding,
    dilations?: [number, number]|number, strides?: [number, number]|number,
    dimRoundingMode?: 'floor'|'round'|'ceil', outputSizes?: [number, number]) {
  if (dilations == null) {
    dilations = [1, 1];
  }
  ...
  util.assert(
      conv_util.eitherStridesOrDilationsAreOne(strides, dilations),
      () => 'Error in pool: Either strides or dilations must be 1. ' +
          `Got strides ${strides} and dilations '${dilations}'`);

  const convInfo = conv_util.computePool2DInfo(
      x4D.shape, windowShape, strides, dilations, pad);
  const dilation: [number, number] =
      [convInfo.dilationHeight, convInfo.dilationWidth];


  let basePadding: number[][];
  if (pad === 'same') {
    basePadding = withSpaceToBatchBasePaddings(
        [convInfo.filterHeight, convInfo.filterWidth], dilation);
  } else {
    basePadding = [[0, 0], [0, 0]];
  }

  const isDilationOne = dilation[0] === 1 && dilation[1] === 1;
  const [adjustedPadding, adjustedCrops] = requiredSpaceToBatchPaddings(
      [convInfo.inHeight, convInfo.inWidth], dilation, basePadding);
  const convertedPad = isDilationOne ? pad : 'valid';
  const convertedX =
      isDilationOne ? x4D : spaceToBatchND(x4D, dilation, adjustedPadding);

  const forwardOp = poolingType === 'avg' ?
      () => avgPool(convertedX, windowShape, strides, convertedPad,
                    dimRoundingMode, outputSizes) :
      () => maxPool(convertedX, windowShape, strides, convertedPad,
                    dimRoundingMode, outputSizes);
...
``` 

So to enable these new two options, we still depend on tfjs implementation. Thanks

**Comment by BruceDai** [comment]
For `outputSizes` options, I referred to output shape calculation method of  pool2d on WebNN-native repo by @mingmingtasd.
@huningxin @mingmingtasd PTAL, thanks.

File: src/nn/utils.ts [hunk 1]
```diff
@@ -200,6 +200,20 @@ export function validateAxes(axes: number[], rank: number): boolean {
...
 }
 
+export function product(array: number[]): number {
+  return array.reduce(
+      (accumulator, currentValue) => accumulator * currentValue);
+}
+
+export function checkShape(actual: number[], expected: number[]): void {
+  for (let i = 0; i < actual.length; ++i) {
+    assert(
```
Comment (mingmingtasd) [inline src/nn/utils.ts:210]: Should you compare the length at first?

File: src/nn/ops/pool2d.ts [hunk 2]
```diff
@@ -59,28 +65,141 @@ export abstract class Pool extends SingleOutputOperation {
...
+    this.outputSizes_ = outputSizes;
   }
 
   inputs(): MLOperand[] {
     return [this.input_];
   }
 
+  calculateShape(
+      inputShape:[number, number, number, number],
+      roundingType?: 'ceil' | 'floor' | 'round'): [number, number] {
```
[the remark this answers is not in the document]
Comment (BruceDai) [reply src/nn/ops/pool2d.ts:89]: Spec of pooling operations doesn't define `round` type, but `outputSizes` may be calculated using `round` which tfjs backend supported.

**Comment by BruceDai** [comment]
@mingmingtasd I updated code following your comments, please take another review, thanks.

**Comment by mingmingtasd** [comment]
> @mingmingtasd I updated code following your comments, please take another review, thanks.

Thanks, LGTM! @BruceDai

**Review by huningxin — APPROVED** [review]
LGTM, thanks @BruceDai

**Pull request merged**